### DISCUSSION TITLE
onDataReceived prints 8 times when receiving a signal

### DISCUSSION THREAD

**Discussion by ValentinOsvaldo** [kenjdavidson/react-native-bluetooth-classic#221] [Q&A] [answered]
My device does an action, this action sends a byte array, which I then transform depending on the length of the array, but the device only sends that array 1 time, and on the device I receive the byte array 8 times.
Sometimes it sends the header (identifier) ​​8 times and then the body (byte array) another 8 times.

I don't understand what is happening, because in the arduino IDE the device only sends the byte array 1 time

![image](https://user-images.githubusercontent.com/69062379/206954107-55a4d099-1d74-47f1-a7a9-cbfdf999f09d.png)

My code: 

Connection:
```js
const resp = await device.connect({
  connectorType: 'rfcomm',
  connectionType: 'binary',
  readTimeout: 0,
});
```

On Data received function:
```js
const onDataReceived = ({ data }: BluetoothDeviceReadEvent) => {
  const arrayBuffer = base64ToArrayBuffer(data);
  // Transform array buffer to int array
  // const arrayByte = new Uint8Array(arrayBuffer);
  
  console.log(
    new Uint8Array(arrayBuffer),
    typeof new Uint8Array(arrayBuffer).byteLength,
  );

  // ...
};
```

**Comment by kenjdavidson** [comment]
You'll need to do a little debugging to determine why the duplicate. I can't do much with custom code unless I have your app code, Arduino code and time to put it together.

Off the top of. My head I can't think of anything, but will try to dive deeper at some point.

I personally don't use the binary connection, but others have it working. I suggest starting in the on data Java code and see why it thinks there are 8 messages.

**Reply by ValentinOsvaldo** [reply]
I think it's a problem related to re-renders, but I don't understand how this influences my onDataReceived function.

I use a ```useEffect``` and put dependency on my connected device, as shown here:

```tsx
useEffect(() => {
  if (!deviceConnected) return;
  deviceConnected.onDataReceived(onDataReceived);
  console.log('onDataReceived');
}, [deviceConnected]);
```

But when I put the listener in the function when connecting the device, it works normal, but really I would like to be able to put my onDataReceived in a ```useEffect``` with some dependencies. I don't know if you would recommend using the  ```useEffect``` really

Sorry if this is getting off topic, but I really have no other idea how to do it...

**Comment by kenjdavidson** [comment, marked as answer]
If you're in use effect with a condition you shouldn't need an if statement. There may be an issue with the way react equals the condition? If you're seeing it set the listener multiple times.

You should also be using a cleanup in use effect for things that need to unsubscribe.

**Comment by ValentinOsvaldo** [comment]
Thank you very much for your help 🙏

Ok, I feel really dumb but I hope this post helps more people, my mistake was using my custom hook in several components. 

Please use your custom hook in a single component

**Reply by kenjdavidson** [reply]
Right on, glad it got sorted out.

**Reply by ValentinOsvaldo** [reply]
Just one more question, now they inform me that they send an array of bytes but a part of the array is received and then the rest. What could it be due to?

![image](https://user-images.githubusercontent.com/69062379/207975058-95a40bf8-9a06-47a9-aa11-7c4c96598197.png)

```[170, 20, 174, 119, 65, 10]``` is the complete array and ```[170, 154]``` and ```[153, 143, 65, 10]``` is another array but it is split

**Comment by kenjdavidson** [comment]
The `binary` connection is done based on chunks, so it's entirely possible that 2 bytes then 4 bytes are written.  If you know your message format is 6 bytes, you'll need to stitch them together on the application side.  If you need to extend the `ByteArrayConnectionImpl` then you can do so, and only send data to the application after X bytes.
